pkg/applicationserver/io/pubsub/grpc_pubsub.go

```diff
@@ -0,0 +1,118 @@
+
+package pubsub
+
+import (
+	"context"
+
+	pbtypes "github.com/gogo/protobuf/types"
+	"go.thethings.network/lorawan-stack/pkg/auth/rights"
+	"go.thethings.network/lorawan-stack/pkg/errors"
+	"go.thethings.network/lorawan-stack/pkg/log"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+)
+
+// appendImplicitPubSubGetPaths appends implicit ttnpb.ApplicationPubSub get paths to paths.
+func appendImplicitPubSubGetPaths(paths ...string) []string {
+	return append(append(make([]string, 0, 3+len(paths)),
+		"format",
+		"provider",
+		"base_topic",
+	), paths...)
+}
+
+// GetFormats implements ttnpb.ApplicationPubSubRegistryServer.
+func (ps *PubSub) GetFormats(ctx context.Context, _ *pbtypes.Empty) (*ttnpb.ApplicationPubSubFormats, error) {
+	fs := make(map[string]string, len(formats))
+	for key, val := range formats {
+		fs[key] = val.Name
+	}
+	return &ttnpb.ApplicationPubSubFormats{
+		Formats: fs,
+	}, nil
+}
+
+// Get implements ttnpb.ApplicationPubSubRegistryServer.
+func (ps *PubSub) Get(ctx context.Context, req *ttnpb.GetApplicationPubSubRequest) (*ttnpb.ApplicationPubSub, error) {
+	if err := rights.RequireApplication(ctx, req.ApplicationIdentifiers, ttnpb.RIGHT_APPLICATION_TRAFFIC_READ); err != nil {
+		return nil, err
+	}
+	return ps.registry.Get(ctx, req.ApplicationPubSubIdentifiers, appendImplicitPubSubGetPaths(req.FieldMask.Paths...))
+}
+
+// List implements ttnpb.ApplicationPubSubRegistryServer.
+func (ps *PubSub) List(ctx context.Context, req *ttnpb.ListApplicationPubSubsRequest) (*ttnpb.ApplicationPubSubs, error) {
+	if err := rights.RequireApplication(ctx, req.ApplicationIdentifiers, ttnpb.RIGHT_APPLICATION_TRAFFIC_READ); err != nil {
+		return nil, err
+	}
+	pubsubs, err := ps.registry.List(ctx, req.ApplicationIdentifiers, appendImplicitPubSubGetPaths(req.FieldMask.Paths...))
+	if err != nil {
+		return nil, err
+	}
+	return &ttnpb.ApplicationPubSubs{
+		Pubsubs: pubsubs,
+	}, nil
+}
+
+// Set implements ttnpb.ApplicationPubSubRegistryServer.
+func (ps *PubSub) Set(ctx context.Context, req *ttnpb.SetApplicationPubSubRequest) (*ttnpb.ApplicationPubSub, error) {
+	if err := rights.RequireApplication(ctx, req.ApplicationIdentifiers, ttnpb.RIGHT_APPLICATION_TRAFFIC_READ); err != nil {
+		return nil, err
+	}
+	// Get all the fields here for starting the integration task.
+	pubsub, err := ps.registry.Set(ctx, req.ApplicationPubSubIdentifiers, appendImplicitPubSubGetPaths(req.FieldMask.Paths...),
+		func(pubsub *ttnpb.ApplicationPubSub) (*ttnpb.ApplicationPubSub, []string, error) {
+			if pubsub != nil {
+				return &req.ApplicationPubSub, req.FieldMask.Paths, nil
+			}
+			return &req.ApplicationPubSub, append(req.FieldMask.Paths,
+				"ids.application_ids",
+				"ids.pub_sub_id",
+			), nil
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+	if err := ps.stop(ctx, req.ApplicationPubSubIdentifiers); err != nil && !errors.IsNotFound(err) {
+		log.FromContext(ctx).WithError(err).Warn("Failed to cancel integration")
```

Comment:
- `... to stop pubsub`
- add relevant fields (e.g. uids) to `logger`

Reply:
The UIDs are added above (line 74).